### PR TITLE
ACL Authorizer overhaul

### PR DESCRIPTION
**Pull request by mkeeler**
Implementation Notes:
- The `Authorizer` interfaces method signatures were updated in two ways. 
   - The methods now take a `*EnterpriseAuthorizerContext` object which will be used for extra validations in Consul enterprise. In the places where we support Sentinel this new struct has taken the place of any Sentinel scope functions as that scope will be set within the context.
   - The methods now return an `acl.EnforcementDecision`. In order to reuse the "chaining" behavior" of the previous `PolicyAuthorizer` type, the chaining behavior was moved out into its own type. `Authorizer` implementors now can return an authoritative `Allow` or `Deny` decisions or in the case where they don't have an answer will return `Default` and allow the next `Authorizer` in the chain to be queried.
- The `Policy` struct was refactored to move all the policy rules lists out into a separate `PolicyRules` struct and that one is now embedded within the `Policy` struct. Additionally an `EnterprisePolicyRules` struct with a stub interface was created to hold Consul Enterprise specific rules. It will become clearer with the corresponding PRs for Consul Enterprise, but this `PolicyRules` struct embedding is to allow for reusing these rules even within enterprise policies.
- For the time being sentinel scope creation is disabled but will be brought back in a subsequent PR. Really all the places where we enforce ACLs we will have to be able to provide some `EnterpriseAuthorizerContext` and for KV writes this will include the sentinel scope. For the sake of keeping this PR as minimal as possible I have left this to the future but there are tons of `TODO (namespaces)` comments indicating where we need to follow on and create the proper authz context.

Testing Notes:

- acl/acl_test.go received the most minimal modifications that I could do in order to ensure that the tests that passed before the refactoring still pass in the same manner. All the `check*` functions can now take a `*EnterpriseAuthorizerContext` although in the OSS codebase tests its always nil. These functions also have to account for the method signature changes to return an `EnforcementDecision` vs a `bool`. Lastly all the inlined policies were updated to account for the `PolicyRules` struct embedded within the `Policy` struct.
- acl/policy_test.go received the most minimal modifications that I could do in order to ensure that the tests that passed before the refactoring still pass in the same manner. This mainly just involved wrapping all the expected policies within a `PolicyRules` struct.

### CODE REVIEW

File: acl/authorizer.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
+package acl
+
+type EnforcementDecision int
+
+const (
+	// Allow returned from an Authorizer enforcement method indicates
+	// that a corresponding rule was found and that access should be allowed
+	Allow EnforcementDecision = iota
```
Comment (rboyer) [inline acl/authorizer.go]: Do you want the default zero value of `EnforcementDecision` to be `Allow`?

Comment (mkeeler) [reply acl/authorizer.go]: Good point. In theory it shouldn't matter but will swap them just in case.

File: acl/policy_authorizer_oss.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+// enterprisePolicyAuthorizer stub
+type enterprisePolicyAuthorizer struct{}
+
+func (authz *enterprisePolicyAuthorizer) init(*EnterpriseACLConfig) {
+	// nothing to do
+}
+
+func (authz *enterprisePolicyAuthorizer) enforce(*EnterpriseRule, *EnterpriseAuthorizerContext) EnforcementDecision {
+	return Allow
```
Comment (rboyer) [inline acl/policy_authorizer_oss.go]: Shouldn't this be `Default` for OSS? It feels weird to not in some way have this layer using the default policy action from config.

Comment (mkeeler) [reply acl/policy_authorizer_oss.go]: I changed it and added a `defaultIsAllow` func to translate any `Allow` or `Default` response to an `Allow`

**Pull request merged**